source/server/listener_manager_impl.cc

```diff
@@ -780,6 +788,25 @@ uint64_t ListenerManagerImpl::numConnections() {
   return num_connections;
 }
 
+void ListenerManagerImpl::stopListener(Network::ListenerConfig& listener) {
+  ENVOY_LOG(debug, "begin stop listener: name={}", listener.name());
+  for (const auto& worker : workers_) {
+    if (listener.direction() == envoy::api::v2::core::TrafficDirection::INBOUND) {
+      inbound_listeners_stopped_ = true;
+    }
+    if (listener.direction() == envoy::api::v2::core::TrafficDirection::OUTBOUND) {
+      outbound_listeners_stopped_ = true;
+    }
```

Comment:
I don't think this works. Notably, if you hit `/drain_listeners` to stop all listeners, and all of them have unspecified traffic direction, then neither `inbound_listeners_stopped_` nor `outbound_listeners_stopped_` will be set to true.

Reply:
The approach as discussed https://github.com/envoyproxy/envoy/issues/4460#issuecomment-535449831 is based on traffic direction being set to distinguish between inbound vs outbound listeners as that is a simple way to distinguish. Otherwise I am not sure of any other way to easily distinguish. Since it is needed for inbound, we might as well use it for` /drain_listeners` - to stop all listeners instead of using a different approach there. IMO, that is a fair assumption and an easy change on control plane side to set the listener direction -Istio already does it for example.

Reply:
The approach is fine (using configured traffic direction when draining inbound/outbound listeners).

However, the implementation isn't, since it's setting the state based on the configured traffic direction of the listener that's being drained, and not on what was requested, e.g. a request to drain all listeners (using `/drain_listeners`), when none of the listeners have configured traffic direction, won't modify `inbound_listeners_stopped_` and `outbound_listeners_stopped_`, and `listnerStopped()` will always return `false`.

Reply:
I see. I will make change to ignore stop listener/ not stop listner if traffic_direction is not set (UNSPECIFIED). Is that fine?

Reply:
No, when draining ALL listeners (`/drain_listeners`), traffic direction shouldn't matter. All listeners should be drained, and no new listeners should be accepted using LDS.

Reply:
OK. Changed it like that. PTAL. I changed the new method to `shutdownListeners` to better describe what we are doing here.